common/cluster/metadata.go

```diff
@@ -212,7 +304,196 @@ func (m *metadataImpl) ClusterNameForFailoverVersion(failoverVersion int64) stri
+			newEntries[clusterName] = &ClusterInformation{
+				Enabled:                newClusterInfo.Enabled,
+				InitialFailoverVersion: newClusterInfo.InitialFailoverVersion,
+				RPCAddress:             newClusterInfo.RPCAddress,
+				version:                newClusterInfo.version,
+			}
+		} else if newClusterInfo.version > oldClusterInfo.version {
+			if newClusterInfo.Enabled == oldClusterInfo.Enabled &&
+				newClusterInfo.RPCAddress == oldClusterInfo.RPCAddress &&
+				newClusterInfo.InitialFailoverVersion == oldClusterInfo.InitialFailoverVersion {
+				// key cluster info does not change
+				continue
+			}
+			// handle updated cluster registry
+			oldEntries[clusterName] = &ClusterInformation{
+				Enabled:                oldClusterInfo.Enabled,
+				InitialFailoverVersion: oldClusterInfo.InitialFailoverVersion,
+				RPCAddress:             oldClusterInfo.RPCAddress,
+				version:                oldClusterInfo.version,
+			}
+			newEntries[clusterName] = &ClusterInformation{
+				Enabled:                newClusterInfo.Enabled,
+				InitialFailoverVersion: newClusterInfo.InitialFailoverVersion,
+				RPCAddress:             newClusterInfo.RPCAddress,
+				version:                newClusterInfo.version,
+			}
+		}
+	}
+	for clusterName, oldClusterInfo := range clusterInfoMap {
+		if _, ok := clusterMetadataMap[clusterName]; !ok {
+			// removed cluster registry
+			oldEntries[clusterName] = &oldClusterInfo
+			newEntries[clusterName] = nil
+		}
+	}
+
+	if len(oldEntries) > 0 {
+		m.clusterLock.Lock()
+		m.updateClusterInfoLocked(oldEntries, newEntries)
+		m.updateFailoverVersionToClusterName()
+		m.clusterLock.Unlock()
+
+		m.clusterCallbackLock.RLock()
+		defer m.clusterCallbackLock.RUnlock()
+		for _, cb := range m.clusterChangeCallback {
+			cb(oldEntries, newEntries)
+		}
+	}
+	return nil
+}
+
+func (m *metadataImpl) updateClusterInfoLocked(
+	oldClusterMetadata map[string]*ClusterInformation,
+	newClusterMetadata map[string]*ClusterInformation,
+) {
+	for clusterName := range oldClusterMetadata {
+		if oldClusterMetadata[clusterName] != nil && newClusterMetadata[clusterName] == nil {
+			delete(m.clusterInfo, clusterName)
+		} else {
+			m.clusterInfo[clusterName] = *newClusterMetadata[clusterName]
+		}
+	}
+}
+
+func (m *metadataImpl) updateFailoverVersionToClusterName() {
+	m.versionToClusterName = updateVersionToClusterName(m.clusterInfo, m.failoverVersionIncrement)
+}
+
+func updateVersionToClusterName(clusterInfo map[string]ClusterInformation, failoverVersionIncrement int64) map[int64]string {
+	versionToClusterName := make(map[int64]string)
+	for clusterName, info := range clusterInfo {
+		if failoverVersionIncrement <= info.InitialFailoverVersion || info.InitialFailoverVersion <= 0 {
+			panic(fmt.Sprintf(
+				"Version increment %v is smaller than initial version: %v.",
+				failoverVersionIncrement,
+				clusterInfo,
+			))
+		}
+		if len(clusterName) == 0 {
+			panic("Cluster name in all cluster names is empty")
```

Comment:
message is confusing. better to include more details (other fields of this clusterInfo etc).

Reply:
updated